x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/repository/CcrRestoreSourceServiceTests.java

```diff
@@ -167,24 +168,43 @@ public void testGetSessionDoesNotLeakFileIfClosed() throws IOException {
             indexDoc(indexShard, "_doc", Integer.toString(i));
             flushShard(indexShard, true);
         }
-        final String sessionUUID1 = UUIDs.randomBase64UUID();
+        final String sessionUUID = UUIDs.randomBase64UUID();
 
-        restoreSourceService.openSession(sessionUUID1, indexShard);
+        restoreSourceService.openSession(sessionUUID, indexShard);
 
         ArrayList<StoreFileMetaData> files = new ArrayList<>();
         indexShard.snapshotStoreMetadata().forEach(files::add);
 
-        try (CcrRestoreSourceService.SessionReader sessionReader = restoreSourceService.getSessionReader(sessionUUID1)) {
+        try (CcrRestoreSourceService.SessionReader sessionReader = restoreSourceService.getSessionReader(sessionUUID)) {
             sessionReader.readFileBytes(files.get(0).name(), new BytesArray(new byte[10]));
         }
 
         // Request a second file to ensure that original file is not leaked
-        try (CcrRestoreSourceService.SessionReader sessionReader = restoreSourceService.getSessionReader(sessionUUID1)) {
+        try (CcrRestoreSourceService.SessionReader sessionReader = restoreSourceService.getSessionReader(sessionUUID)) {
             sessionReader.readFileBytes(files.get(1).name(), new BytesArray(new byte[10]));
         }
 
-        restoreSourceService.closeSession(sessionUUID1);
+        restoreSourceService.closeSession(sessionUUID);
         closeShards(indexShard);
         // Exception will be thrown if file is not closed.
     }
+
+    public void testSessionCanTimeout() throws Exception {
+        restoreSourceService = new CcrRestoreSourceService(Settings.EMPTY, threadPool, new TimeValue(10, TimeUnit.MILLISECONDS));
+        IndexShard indexShard = newStartedShard(true);
+
+        final String sessionUUID = UUIDs.randomBase64UUID();
+
+        assertBusy(() -> {
+            restoreSourceService.openSession(sessionUUID, indexShard);
+            LockSupport.parkNanos(TimeUnit.MILLISECONDS.toNanos(100));
```

Comment:
timing-dependent tests are evil and can easily fail on CI. We have an alternative way to write these tests, so that they are fully reproducible. `DeterministicTaskQueue` gives you a fake `ThreadPool` that allows for full mocking of timing. See `DeterministicTaskQueueTests.testThreadPoolSchedulesPeriodicFutureTasks` on how to use this.